src/VisualStudio/Core/Def/Implementation/SolutionSize/SolutionSizeTracker.cs

```diff
@@ -57,79 +56,29 @@ public Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancel
                 return SpecializedTasks.EmptyTask;
             }
 
-            public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
+            public async Task AnalyzeSyntaxAsync(Document document, CancellationToken cancellationToken)
             {
-                var sum = await GetProjectSizeAsync(project, cancellationToken).ConfigureAwait(false);
+                // getting tree is cheap since tree always stays in memory
+                var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
 
-                _map.AddOrUpdate(project.Id, sum, (id, existing) => sum);
-
-                _size = _map.Values.Sum();
-            }
-
-            public void RemoveProject(ProjectId projectId)
-            {
-                long unused;
-                _map.TryRemove(projectId, out unused);
-
-                _size = _map.Values.Sum();
-            }
-
-            private static async Task<long> GetProjectSizeAsync(Project project, CancellationToken cancellationToken)
-            {
-                if (project == null)
-                {
-                    return 0;
-                }
-
-                var sum = 0L;
-                foreach (var document in project.Documents)
+                // remove old size
+                long size;
+                if (_map.TryGetValue(document.Id, out size))
                 {
-                    sum += await GetDocumentSizeAsync(document, cancellationToken).ConfigureAwait(false);
+                    _size -= size;
```

Comment:
If not (guaranteed to run sequentially), then do we actually need ```ConcurrentDictionary``` or would a regular ```Dictionary<K,V>``` suffice?